iree/compiler/Dialect/Flow/Transforms/UnrollReductions.cpp

```diff
@@ -0,0 +1,180 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "iree/compiler/Dialect/Flow/Transforms/Passes.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/Pass/Pass.h"
+#include "tensorflow/compiler/mlir/xla/ir/hlo_ops.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+namespace {
+
+// Unrolls a multi-dimensional xla_hlo.reduce op into one xla_hlo.reduce op per
+// dimension. The XLA operation semantics state that this is a valid
+// transformation.
+void unrollReduceOp(xla_hlo::ReduceOp reduceOp) {
+  // Create one op per dimension being reduced.
+  // We'll do this by chaining the original input through with the temporary
+  // reduction results. The results we end up with will be the originally
+  // requested shape and we can just substitute them.
+  SmallVector<int64_t, 4> sortedDimensions{
+      reduceOp.dimensions().getValues<int64_t>()};
+  llvm::sort(sortedDimensions,
+             [](int64_t a, int64_t b) { return (a - b) > 0; });
+
+  // Insert at the same place as the original op.
+  OpBuilder builder(reduceOp);
+  SmallVector<Value, 4> temps{reduceOp.operands()};
+  for (int64_t dimension : sortedDimensions) {
+    // Create the new reduction using the results of the previous operation.
+    auto singleAttrType =
+        RankedTensorType::get({1}, builder.getIntegerType(64));
+    auto singleReduceOp = builder.create<xla_hlo::ReduceOp>(
+        reduceOp.getLoc(), temps, reduceOp.init_values(),
+        DenseIntElementsAttr::get(
+            singleAttrType,
+            {builder.getI64IntegerAttr(dimension).cast<Attribute>()}));
```

Comment:
You should just be able to pass an int here, i.e. replace all this with dimension. Same below